File: app/src/main/java/vergecurrency/vergewallet/wallet/Credentials.java
package vergecurrency.vergewallet.wallet;


import java.security.MessageDigest;
import java.util.Arrays;
import java.util.Base64;
import java.util.List;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

import io.horizontalsystems.hdwalletkit.HDKey;
import io.horizontalsystems.hdwalletkit.HDKeychain;
import io.horizontalsystems.hdwalletkit.HDPublicKey;
import kotlin.text.Charsets;
import vergecurrency.vergewallet.service.model.MnemonicManager;
import vergecurrency.vergewallet.service.model.PreferencesManager;

public class Credentials {

	private HDKeychain hdKeychain;
	public Credentials() {
		hdKeychain = initHDKeyChain();
	}
	public HDKeychain initHDKeyChain() {
		String[] mnemonic = MnemonicManager.getMnemonicFromJSON(PreferencesManager.getMnemonic());
		String passphrase = PreferencesManager.getPassphrase();

		byte[] seed = new io.horizontalsystems.hdwalletkit.Mnemonic().toSeed((List<String>) Arrays.asList(mnemonic), passphrase);
		return new HDKeychain(seed, true);
	}

	public HDKey getWalletPrivateKey() {
		//hardened at 0
		return hdKeychain.getKeyByPath("/0'");
	}

	public HDKey getRequestWalletPrivateKey() {
		return hdKeychain.getKeyByPath("/1'/0");
	}


	public HDKey getBip44PrivateKey() {
		return hdKeychain.getKeyByPath("m/44'/0'/0'");
	}

	public HDPublicKey getHDPublicKey() {
		return new HDPublicKey(0, false, getBip44PrivateKey());
	}

	public String getPersonalEncryptingKey() {
		try {
			byte[] data = MessageDigest.getInstance("SHA-256").digest(getRequestWalletPrivateKey().getPrivKeyBytes());
			byte[] key = "personalKey".getBytes(Charsets.UTF_8);

			Mac mac = Mac.getInstance("HmacSHA256");
			mac.init(new SecretKeySpec(key, "HmacSHA256"));
			return Base64.getEncoder().encodeToString(Arrays.copyOf(mac.doFinal(data), 16));
		} catch (Exception e) {
			e.printStackTrace();
			return null;
		}
	}

	public String getSharedEncryptingKey() {
		try {
			byte [] data = MessageDigest.getInstance("SHA-256").digest(getWalletPrivateKey().getPrivKeyBytes());
			return Base64.getEncoder().encodeToString(Arrays.copyOf(data, 16));
		} catch (Exception e) {
			e.printStackTrace();
			return null;
		}
	}
}

File: app/src/main/kotlin/vergecurrency/vergewallet/wallet/Credentials.kt
package vergecurrency.vergewallet.wallet


import io.horizontalsystems.bitcoinkit.BitcoinKit
import io.horizontalsystems.hdwalletkit.*
import vergecurrency.vergewallet.service.model.MnemonicManager
import vergecurrency.vergewallet.service.model.PreferencesManager
import java.security.MessageDigest
import java.security.PrivateKey
import java.util.*
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

class Credentials(mnemonic: Array<String>, passphrase: String, networkType: BitcoinKit.NetworkType = BitcoinKit.NetworkType.MainNet) {

    var seed: ByteArray = Mnemonic().toSeed(mnemonic.toList(), passphrase)
    var network: BitcoinKit.NetworkType = networkType


    private val privateKey: HDKey
    get() = HDKeyDerivation.createRootKey(seed)

    //hardened at 0
    val walletPrivateKey: HDKey
        get() = HDKeyDerivation.deriveChildKey(privateKey,0,true)

    val requestPrivateKey: HDKey
        get() = HDKeyDerivation.deriveChildKey(HDKeyDerivation.deriveChildKey(privateKey,1,true),0,false)


    val bip44PrivateKey: HDKey
        get() = HDKeyDerivation.deriveChildKey(HDKeyDerivation.deriveChildKey(HDKeyDerivation.deriveChildKey(privateKey,44,true),0,true),0,true)

    val publicKey: HDPublicKey
        get() = HDPublicKey(0, false, bip44PrivateKey)


    val personalEncryptingKey: String?
        get() {
            try {
                val data = MessageDigest.getInstance("SHA-256").digest(requestPrivateKey.privKeyBytes)
                val key = "personalKey".toByteArray(Charsets.UTF_8)
                val mac = Mac.getInstance("HmacSHA256")
                mac.init(SecretKeySpec(key, "HmacSHA256"))
                return Base64.getEncoder().encodeToString(Arrays.copyOf(mac.doFinal(data), 16))
            } catch (e: Exception) {
                e.printStackTrace()
                return null
            }

        }

    val sharedEncryptingKey: String?
        get() {
            try {
                val data = MessageDigest.getInstance("SHA-256").digest(walletPrivateKey.privKeyBytes)
                return Base64.getEncoder().encodeToString(Arrays.copyOf(data, 16))
            } catch (e: Exception) {
                e.printStackTrace()
                return null
            }

        }


    fun privateKeyBy(path: String, privateKey: HDKey) :ByteArray {
        var key = privateKey

        for (deriver in path.replace("m/", "").split("/")) {
            try {
                var deriverInt = Integer.parseInt(deriver)
                key = HDKeyDerivation.deriveChildKey(key,deriverInt, false)

            } catch (e: NumberFormatException) {
                throw Exception("Invalid deriver exception")
            }
        }
        return key.privKeyBytes
    }

}
